core/src/qos_service.rs

```diff
@@ -93,28 +106,91 @@ impl QosService {
             .map(|(tx, cost)| match cost_tracker.try_add(tx, cost) {
                 Ok(current_block_cost) => {
                     debug!("slot {:?}, transaction {:?}, cost {:?}, fit into current block, current block cost {}", bank.slot(), tx, cost, current_block_cost);
-                    stats.selected_txs_count += 1;
+                    self.metrics.selected_txs_count.fetch_add(1, Ordering::Relaxed);
                     Ok(())
                 },
                 Err(e) => {
                     debug!("slot {:?}, transaction {:?}, cost {:?}, not fit into current block, '{:?}'", bank.slot(), tx, cost, e);
                     match e {
                         CostTrackerError::WouldExceedBlockMaxLimit => {
-                            stats.retried_txs_per_block_limit_count += 1;
+                            self.metrics.retried_txs_per_block_limit_count.fetch_add(1, Ordering::Relaxed);
                             Err(TransactionError::WouldExceedMaxBlockCostLimit)
                         }
                         CostTrackerError::WouldExceedAccountMaxLimit => {
-                            stats.retried_txs_per_account_limit_count += 1;
+                            self.metrics.retried_txs_per_account_limit_count.fetch_add(1, Ordering::Relaxed);
                             Err(TransactionError::WouldExceedMaxAccountCostLimit)
                         }
                     }
                 }
             })
             .collect();
         cost_tracking_time.stop();
-        stats.cost_tracking_time += cost_tracking_time.as_us();
+        self.metrics
+            .cost_tracking_time
+            .fetch_add(cost_tracking_time.as_us(), Ordering::Relaxed);
         select_results
     }
+
+    fn reporting_loop(running_flag: Arc<AtomicBool>, metrics: Arc<QosServiceMetrics>) {
+        while running_flag.load(Ordering::Relaxed) {
+            // hardcode to report every 1000ms
```

Comment:
Will we ever want this tunable?

Reply:
Yeah, initially I thought only care accumulated numbers; but as going through last few stall instigations, I wished they were by slots. So another change is coming soon to report per slot